src/mbgl/renderer/render_layer.hpp

```diff
@@ -58,6 +66,8 @@ class RenderLayer {
     // Checks whether this layer can be rendered.
     bool needsRendering(float zoom) const;
 
+    virtual void uploadBuckets(gl::Context&, RenderSource* source);
+    virtual void render(Painter& , PaintParameters& , RenderSource* source);
```

Comment:
These will be overridden in RenderRasterLayer to support ImageSource. The raster layer will determine the correct upload/render method based on the source type.